README.md
<!--
 Copyright 2023 Dimitri Koshkin. All rights reserved.
 SPDX-License-Identifier: Apache-2.0
 -->

# kubectl-assistant

This tool uses [OpenAI's GPT3](https://platform.openai.com/account/api-keys) API to generate Kubernetes [kubectl](https://kubernetes.io/docs/reference/kubectl/) commands to run against a cluster.
It aims to help find `kubectl` commands for what you are trying to accomplish, without needing to search around in the docs.

## Prerequisites

- A running Kubernetes cluster
- [kubectl](https://kubernetes.io/docs/reference/kubectl/) in your `PATH`

## Usage Instructions

1.  Download the binary for your OS and architecture.

2.  Move it somewhere in your `PATH`.

3.  Generate an OpenAI API Key [here](https://platform.openai.com/account/api-keys).

4.  Run `kubectl assistant`

    ```txt
    $ export OPENAI_KEY=<>
    $ kubectl assistant
    Begin by typing what you want to accomplish in your Kubernetes cluster and then hit "Enter".
    For example:
      List all control-plane Nodes
      Get Kubernetes versions for all Nodes
      Create deployment named nginx, using image nginx and ports 80 and 443
      Find all objects with label app=nginx

    You will then see some text output and in most cases either a exec command or some YAML output.
    If the command looks reasonable to you, type in "k" and then hit "Enter" to execute it against the cluster.

    You can also type "kubectl ..." to execute a custom command.

    Hit CTRL+C to exit.

    > List all control-plane Nodes
    ==============================================================================================================================================================================================================
    To list all control-plane nodes in your Kubernetes cluster, you can use the following exec command:

    ```bash
    kubectl get nodes --selector=node-role.kubernetes.io/control-plane
    ``

    This command will display a list of all the control-plane nodes in the cluster. The `--selector=node-role.kubernetes.io/control-plane` flag filters the list of nodes based on a label selector that is automatically applied to control-plane nodes by default when they are registered with the cluster.

    > k
    NAME                 STATUS   ROLES           AGE     VERSION
    kind-control-plane   Ready    control-plane   3d16h   v1.25.3

    > kubectl get pods -A
    NAMESPACE            NAME                                         READY   STATUS    RESTARTS   AGE
    default              nginx                                        1/1     Running   0          3d16h
    default              nginx-56fd7f4d49-v67dd                       1/1     Running   0          3d15h
    kube-system          coredns-565d847f94-kngc2                     1/1     Running   0          3d16h
    kube-system          coredns-565d847f94-ngfzd                     1/1     Running   0          3d16h
    kube-system          etcd-kind-control-plane                      1/1     Running   0          3d16h
    kube-system          kindnet-t69vm                                1/1     Running   0          3d16h
    kube-system          kube-apiserver-kind-control-plane            1/1     Running   0          3d16h
    kube-system          kube-controller-manager-kind-control-plane   1/1     Running   0          3d16h
    kube-system          kube-proxy-6mb6r                             1/1     Running   0          3d16h
    kube-system          kube-scheduler-kind-control-plane            1/1     Running   0          3d16h
    local-path-storage   local-path-provisioner-684f458cdd-tscq6      1/1     Running   0          3d16h
    ```

After typing in a question in the prompt you will see a response, many of them having example `kubectl ...` commands.

-   Type in `k` to execute the first command against your Kubernetes cluster.

-   Type in another question to get additional responses. (NOTE, unlike ChatGPT the conversation threads are not contextual, if you want more detail for a previois response, you must re-ask the question).

-   Type in `kubectl ...` commands directly to have them execute against the cluster.

## Setup your Dev Environment

- Install [asdf](https://asdf-vm.com/)
- Install [asdf-direnv](https://github.com/asdf-community/asdf-direnv#setup)
- Add a global `direnv` version with: `asdf global direnv latest`
- Install all tools with: `make install-tools`

Tip: to see all available make targets with descriptions, simply run `make`.

### Lint

```bash
make lint
```

### Test

```bash
make test
```

### Build

The binary for your OS will be placed in `./dist`, e.g. `./dist/kubectl-assistant_darwin_arm64/kubectl-assisant`:

```bash
make build-snapshot
```

### Pre-commit

```bash
make pre-commit
```


CHANGELOG.md
# Changelog

<!-- Release notes generated using configuration in .github/release.yaml at main -->

## What's Changed
### Other Changes
* ci: Run actionlint in CI by @dkoshkin in https://github.com/dkoshkin/kubectl-assistant/pull/27
* build: Cleanup tooling and dev docs by @dkoshkin in https://github.com/dkoshkin/kubectl-assistant/pull/29


**Full Changelog**: https://github.com/dkoshkin/kubectl-assistant/compare/v0.1.5...v0.1.6

## Changelog

<!-- Release notes generated using configuration in .github/release.yaml at main -->

## What's Changed
### Other Changes
* ci: Fix asdf setup in Github Actions by @dkoshkin in https://github.com/dkoshkin/kubectl-assistant/pull/25


**Full Changelog**: https://github.com/dkoshkin/kubectl-assistant/compare/v0.1.4...v0.1.5

## Changelog

<!-- Release notes generated using configuration in .github/release.yaml at main -->



**Full Changelog**: https://github.com/dkoshkin/kubectl-assistant/compare/v0.1.3...v0.1.4

## Changelog

<!-- Release notes generated using configuration in .github/release.yaml at main -->

## What's Changed
### Fixes 🔧
* fix: Handle snippets without a newline by @dkoshkin in https://github.com/dkoshkin/kubectl-assistant/pull/20


**Full Changelog**: https://github.com/dkoshkin/kubectl-assistant/compare/v0.1.2...v0.1.3

## Changelog

<!-- Release notes generated using configuration in .github/release.yaml at main -->

## What's Changed
### Fixes 🔧
* fix: Use a more descriptive OpenAI API_KEY env by @dkoshkin in https://github.com/dkoshkin/kubectl-assistant/pull/17


**Full Changelog**: https://github.com/dkoshkin/kubectl-assistant/compare/v0.1.1...v0.1.2

## Changelog

<!-- Release notes generated using configuration in .github/release.yaml at main -->

## What's Changed
### Fixes 🔧
* fix: Add dev docs by @dkoshkin in https://github.com/dkoshkin/kubectl-assistant/pull/14
* fix: Better examples that work by @dkoshkin in https://github.com/dkoshkin/kubectl-assistant/pull/13
* fix: Handle additional snippets by @dkoshkin in https://github.com/dkoshkin/kubectl-assistant/pull/12
* fix: README snippet by @dkoshkin in https://github.com/dkoshkin/kubectl-assistant/pull/16


**Full Changelog**: https://github.com/dkoshkin/kubectl-assistant/compare/v0.1.0...v0.1.1

## Changelog

<!-- Release notes generated using configuration in .github/release.yaml at main -->



**Full Changelog**: https://github.com/dkoshkin/kubectl-assistant/commits/v0.1.0


pkg/prompter/prompter_test.go
// Copyright 2023 Dimitri Koshkin. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package prompter

import (
	"bytes"
	"context"
	"strings"
	"testing"

	"github.com/golang/mock/gomock"
	"github.com/stretchr/testify/assert"

	"github.com/dkoshkin/kubectl-assistant/pkg/assistant"
	"github.com/dkoshkin/kubectl-assistant/pkg/exec"
)

func TestPrompter_Loop_runKubectlDirectly(t *testing.T) {
	ctrl := gomock.NewController(t)

	in := strings.NewReader(`
kubectl get nodes
`)

	mockAssistantRunner := assistant.NewMockRunner(ctrl)
	mockExecRunner := exec.NewMockRunner(ctrl)

	mockExecRunner.EXPECT().IsKubectlCommand("").Return(false).Times(1)
	mockExecRunner.EXPECT().IsKubectlCommand("kubectl get nodes").Return(true).Times(1)
	mockExecRunner.EXPECT().Run("kubectl get nodes").Return(nil).Times(1)

	out := bytes.Buffer{}
	expectedOutput := []byte(
		//nolint:lll // Long lines are fine in tests
		"Begin by typing what you want to accomplish in your Kubernetes cluster and then hit \"Enter\".\nFor example:\n  List all control-plane Nodes\n  Get Kubernetes versions for all Nodes\n  Create deployment named nginx, using image nginx and ports 80 and 443\n  Find all objects with label app=nginx\n\nYou will then see some text output and in most cases either a exec command or some YAML output.\nIf the command looks reasonable to you, type in \"k\" and then hit \"Enter\" to execute it against the cluster.\n\nYou can also type \"kubectl ...\" to execute a custom command.\n\nHit CTRL+C to exit.\n\n> \n> \n> ",
	)

	p := New(mockAssistantRunner, mockExecRunner, in, &out)
	err := p.Loop()
	assert.EqualError(t, err, "an error occurred reading input: EOF", "expected an EOF error")
	assert.Equal(t, string(expectedOutput), out.String())
}

func TestPrompter_Loop_runPromptOutput(t *testing.T) {
	ctrl := gomock.NewController(t)

	in := strings.NewReader(`
list all Nodes
k
`)

	mockAssistantRunner := assistant.NewMockRunner(ctrl)
	mockExecRunner := exec.NewMockRunner(ctrl)

	// expect a call to exec.Run("kubectl get nodes") in assistant.RunKubectlCommand() is called
	mockAssistantRunner.EXPECT().RunKubectlCommand().DoAndReturn(func() error {
		//nolint:wrapcheck // Want to return the error as is in the mock.
		return mockExecRunner.Run("kubectl get nodes")
	})
	mockAssistantRunner.EXPECT().
		GetResponse(context.TODO(), "list all Nodes").
		Return("To list all nodes, run the command:\n\n```\nkubectl get nodes\n```", nil).
		Times(1)
	mockAssistantRunner.EXPECT().CanRunKubectlCommand().Return(true).Times(1)

	mockExecRunner.EXPECT().IsKubectlCommand("").Return(false).Times(1)
	mockExecRunner.EXPECT().IsKubectlCommand("list all Nodes").Return(false).Times(1)
	mockExecRunner.EXPECT().IsKubectlCommand("k").Return(false).Times(1)
	mockExecRunner.EXPECT().Run("kubectl get nodes").Return(nil).Times(1)

	out := bytes.Buffer{}
	expectedOutput := []byte(
		//nolint:lll // Long lines are fine in tests
		"Begin by typing what you want to accomplish in your Kubernetes cluster and then hit \"Enter\".\nFor example:\n  List all control-plane Nodes\n  Get Kubernetes versions for all Nodes\n  Create deployment named nginx, using image nginx and ports 80 and 443\n  Find all objects with label app=nginx\n\nYou will then see some text output and in most cases either a exec command or some YAML output.\nIf the command looks reasonable to you, type in \"k\" and then hit \"Enter\" to execute it against the cluster.\n\nYou can also type \"kubectl ...\" to execute a custom command.\n\nHit CTRL+C to exit.\n\n> \n> ====================================================================================================\nTo list all nodes, run the command:\n\n```\nkubectl get nodes\n```\n\n> \n> ",
	)

	p := New(mockAssistantRunner, mockExecRunner, in, &out)
	err := p.Loop()
	assert.EqualError(t, err, "an error occurred reading input: EOF", "expected an EOF error")
	assert.Equal(t, string(expectedOutput), out.String())
}


pkg/assistant/assistant_test.go
// Copyright 2023 Dimitri Koshkin. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package assistant

import (
	"fmt"
	"testing"

	"github.com/stretchr/testify/assert"
)

func Test_findCodeSnippet(t *testing.T) {
	tests := []struct {
		name            string
		response        string
		expectedCommand string
		expectedErr     error
	}{
		{
			name:            "response with command",
			response:        "To list all nodes, run the command:\n\n```\nkubectl get nodes\n```",
			expectedCommand: "kubectl get nodes",
		},
		{
			name:            "response with command wrapped in ```bash",
			response:        "To list all nodes, run the command:\n\n```bash\nkubectl get nodes\n```",
			expectedCommand: "kubectl get nodes",
		},
		{
			name: "response with command wrapped in ``` without a newline",
			//nolint:lll // Long lines are fine in tests
			response:        "To show deployments with less than desired pods ready, you can use the following command:\n\n```kubectl get deployments --field-selector=\"status.replicas!=status.readyReplicas\"```",
			expectedCommand: "kubectl get deployments --field-selector=\"status.replicas!=status.readyReplicas\"",
		},
		{
			name: "response without a command",
			//nolint:lll // Long lines are fine in tests
			response: "Kubernetes is an open-source container orchestration platform that automates the deployment, scaling, and management of containerized applications. It enables operators to manage applications across multiple nodes, and provides mechanisms for the automated deployment, scaling, and recovery of instances of those applications.",
			// TODO: don't use dynamic errors.
			//nolint:goerr113 // No need to return a custom error.
			expectedErr: fmt.Errorf(
				"did not find opening string in previous output, looking for: ```bash, ```shell, ```",
			),
		},
	}
	for _, tt := range tests {
		gotCommand, gotErr := findCodeSnippet(tt.response)
		assert.Equal(t, tt.expectedErr, gotErr)
		assert.Equal(t, tt.expectedCommand, gotCommand)
	}
}


main.go
// Copyright 2023 Dimitri Koshkin. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package main

import (
	"fmt"
	"io"
	"log"
	"os"
	"os/signal"
	"syscall"

	"github.com/dkoshkin/kubectl-assistant/pkg/assistant"
	"github.com/dkoshkin/kubectl-assistant/pkg/exec"
	"github.com/dkoshkin/kubectl-assistant/pkg/prompter"
	"github.com/dkoshkin/kubectl-assistant/pkg/version"
)

const (
	apiKeyEnv = "OPENAI_KEY"
)

func main() {
	if len(os.Args) > 1 &&
		(os.Args[1] == "version" || os.Args[1] == "-v" || os.Args[1] == "--version") {
		fmt.Println(version.Print())
		os.Exit(0)
	}

	executor := exec.New(io.Discard, os.Stderr)
	if err := executor.KubernetesClusterRunning(); err != nil {
		log.Fatalln("Is a Kubernetes cluster running?")
	}

	apiKey := os.Getenv(apiKeyEnv)
	if apiKey == "" {
		log.Fatalln(
			"OPENAI_KEY with OpenAI key is not set, visit https://platform.openai.com/account/api-keys to create one.",
		)
	}

	sigs := make(chan os.Signal, 1)
	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM, os.Interrupt)
	// Wait for exit signal
	go func() {
		<-sigs
		fmt.Println("Exiting...")
		os.Exit(0)
	}()

	executor = exec.New(os.Stdout, os.Stderr)
	peggy := assistant.New(apiKey, executor)
	ron := prompter.New(peggy, executor, os.Stdin, os.Stdout)

	if err := ron.Loop(); err != nil {
		log.Fatalln(err)
	}
}


pkg/exec/exec.go
// Copyright 2023 Dimitri Koshkin. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package exec

import (
	"io"
	"os/exec"
	"strings"
)

type Runner interface {
	KubernetesClusterRunning() error
	Run(command string) error
	IsKubectlCommand(in string) bool
}

type BashRunner struct {
	out io.Writer
	err io.Writer
}

func New(out, err io.Writer) Runner {
	return BashRunner{out: out, err: err}
}

func (r BashRunner) KubernetesClusterRunning() error {
	return r.Run("kubectl cluster-info")
}

// Run will run the command with 'bash -c'.
func (r BashRunner) Run(command string) error {
	cmd := exec.Command("bash", "-c", command)
	cmd.Stdout = r.out
	cmd.Stderr = r.err
	//nolint:wrapcheck // Want to return the error as is.
	return cmd.Run()
}

// IsKubectlCommand returns true if in starts with "kubectl "
// TODO: make this check more robust.
func (r BashRunner) IsKubectlCommand(in string) bool {
	return strings.HasPrefix(in, "kubectl ")
}


pkg/version/info.go
// Copyright 2023 Dimitri Koshkin. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package version

import (
	"bytes"
	"html/template"
	"runtime"
	"strings"
)

// Build information. Populated at build-time.
//
//nolint:gochecknoglobals // Version globals are set at build time.
var (
	version      string
	major        string
	minor        string
	patch        string
	revision     string
	branch       string
	commitDate   string
	gitTreeState string
	goVersion    = runtime.Version()
)

// Print returns version information.
func Print() string {
	m := map[string]string{
		"version":      version,
		"major":        major,
		"minor":        minor,
		"patch":        patch,
		"revision":     revision,
		"branch":       branch,
		"commitDate":   commitDate,
		"gitTreeState": gitTreeState,
		"goVersion":    goVersion,
		"platform":     runtime.GOOS + "/" + runtime.GOARCH,
	}
	t := template.Must(template.New("version").Parse(`
	version {{.version}} (branch: {{.branch}}, revision: {{.revision}}{{with .gitTreeState}}, gitTreeState: {{.}}{{end}})
		build date:       {{.commitDate}}
		go version:       {{.goVersion}}
		platform:         {{.platform}}
	`))

	var buf bytes.Buffer
	if err := t.ExecuteTemplate(&buf, "version", m); err != nil {
		panic(err)
	}
	return strings.TrimSpace(buf.String())
}


pkg/assistant/mock.go
// Copyright 2023 Dimitri Koshkin. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
//

// Code generated by MockGen. DO NOT EDIT.
// Source: ./pkg/assistant/assistant.go

// Package assistant is a generated GoMock package.
package assistant

import (
	context "context"
	reflect "reflect"

	gomock "github.com/golang/mock/gomock"
)

// MockRunner is a mock of Runner interface.
type MockRunner struct {
	ctrl     *gomock.Controller
	recorder *MockRunnerMockRecorder
}

// MockRunnerMockRecorder is the mock recorder for MockRunner.
type MockRunnerMockRecorder struct {
	mock *MockRunner
}

// NewMockRunner creates a new mock instance.
func NewMockRunner(ctrl *gomock.Controller) *MockRunner {
	mock := &MockRunner{ctrl: ctrl}
	mock.recorder = &MockRunnerMockRecorder{mock}
	return mock
}

// EXPECT returns an object that allows the caller to indicate expected use.
func (m *MockRunner) EXPECT() *MockRunnerMockRecorder {
	return m.recorder
}

// CanRunKubectlCommand mocks base method.
func (m *MockRunner) CanRunKubectlCommand() bool {
	m.ctrl.T.Helper()
	ret := m.ctrl.Call(m, "CanRunKubectlCommand")
	ret0, _ := ret[0].(bool)
	return ret0
}

// CanRunKubectlCommand indicates an expected call of CanRunKubectlCommand.
func (mr *MockRunnerMockRecorder) CanRunKubectlCommand() *gomock.Call {
	mr.mock.ctrl.T.Helper()
	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "CanRunKubectlCommand", reflect.TypeOf((*MockRunner)(nil).CanRunKubectlCommand))
}

// GetResponse mocks base method.
func (m *MockRunner) GetResponse(ctx context.Context, prompt string) (string, error) {
	m.ctrl.T.Helper()
	ret := m.ctrl.Call(m, "GetResponse", ctx, prompt)
	ret0, _ := ret[0].(string)
	ret1, _ := ret[1].(error)
	return ret0, ret1
}

// GetResponse indicates an expected call of GetResponse.
func (mr *MockRunnerMockRecorder) GetResponse(ctx, prompt interface{}) *gomock.Call {
	mr.mock.ctrl.T.Helper()
	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "GetResponse", reflect.TypeOf((*MockRunner)(nil).GetResponse), ctx, prompt)
}

// RunKubectlCommand mocks base method.
func (m *MockRunner) RunKubectlCommand() error {
	m.ctrl.T.Helper()
	ret := m.ctrl.Call(m, "RunKubectlCommand")
	ret0, _ := ret[0].(error)
	return ret0
}

// RunKubectlCommand indicates an expected call of RunKubectlCommand.
func (mr *MockRunnerMockRecorder) RunKubectlCommand() *gomock.Call {
	mr.mock.ctrl.T.Helper()
	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "RunKubectlCommand", reflect.TypeOf((*MockRunner)(nil).RunKubectlCommand))
}


pkg/assistant/assistant.go
// Copyright 2023 Dimitri Koshkin. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package assistant

import (
	"context"
	"errors"
	"fmt"
	"strings"

	"github.com/PullRequestInc/go-gpt3"

	"github.com/dkoshkin/kubectl-assistant/pkg/exec"
)

const (
	//nolint:lll // Long line is fine here
	systemPrompt = "You should return kubectl commands and minimal explanation, do not include sample output. All code snippets should be surrounded by ```"
)

type Runner interface {
	GetResponse(ctx context.Context, prompt string) (string, error)
	CanRunKubectlCommand() bool
	RunKubectlCommand() error
}

type GPT3Runner struct {
	gpt3Client gpt3.Client
	execRunner exec.Runner

	lastResponse string
}

func New(gpt3APIKey string, runner exec.Runner) Runner {
	return &GPT3Runner{
		gpt3Client: gpt3.NewClient(gpt3APIKey),
		execRunner: runner,
	}
}

func (a *GPT3Runner) GetResponse(ctx context.Context, prompt string) (string, error) {
	resp, err := a.gpt3Client.ChatCompletion(ctx, gpt3.ChatCompletionRequest{
		Messages: []gpt3.ChatCompletionRequestMessage{
			{
				Role:    "system",
				Content: systemPrompt,
			},
			{
				Role:    "user",
				Content: prompt,
			},
		},
	})
	if err != nil {
		return "", fmt.Errorf("could not get response from OpenAI: %w", err)
	}

	a.lastResponse = resp.Choices[0].Message.Content

	return a.lastResponse, nil
}

// CanRunKubectlCommand checks if the last response had a exec command.
func (a *GPT3Runner) CanRunKubectlCommand() bool {
	// ignore the error we only care about the command
	command, _ := a.getKubectlCommand()
	return len(command) > 0
}

// RunKubectlCommand will run the exec command from the previous response.
func (a *GPT3Runner) RunKubectlCommand() error {
	command, err := a.getKubectlCommand()
	if err != nil {
		return err
	}
	//nolint:wrapcheck // Want to return the error as is.
	return a.execRunner.Run(command)
}

// getKubectlCommand will return the kubectl command from a code snippet (wrapped with ```).
func (a *GPT3Runner) getKubectlCommand() (string, error) {
	command, err := findCodeSnippet(a.lastResponse)
	if err != nil {
		return "", fmt.Errorf("could not parse last response for a exec command: %w", err)
	}

	if !a.execRunner.IsKubectlCommand(command) {
		//nolint:goerr113 // No need to return a custom error.
		return "", errors.New("command doesn't appear to be a exec command")
	}

	return command, nil
}

func findCodeSnippet(response string) (string, error) {
	start, err := findSnippetOpening(response)
	if err != nil {
		return "", err
	}
	end, err := findSnippetClosing(response[start:])
	if err != nil {
		return "", err
	}

	command := strings.TrimSpace(response[start : start+end])

	return command, nil
}

func findSnippetOpening(str string) (int, error) {
	validStarts := []string{"```bash", "```shell", "```"}

	start := -1
	for _, substr := range validStarts {
		start = strings.Index(str, substr)
		if start != -1 {
			start += len(substr)
			break
		}
	}
	if start == -1 {
		//nolint:goerr113 // No need to return a custom error.
		return start, fmt.Errorf(
			"did not find opening string in previous output, looking for: %s",
			strings.Join(validStarts, ", "),
		)
	}

	return start, nil
}

func findSnippetClosing(str string) (int, error) {
	end := strings.Index(str, "```")
	//nolint:goerr113 // No need to return a custom error.
	if end == -1 {
		return end, fmt.Errorf("did not find closing ``` in previous output")
	}
	// exclude a newline character
	if str[end] == '\n' {
		end--
	}

	return end, nil
}


pkg/prompter/prompter.go
// Copyright 2023 Dimitri Koshkin. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package prompter

import (
	"bufio"
	"context"
	"fmt"
	"io"
	"strings"

	"golang.org/x/term"

	"github.com/dkoshkin/kubectl-assistant/pkg/assistant"
	"github.com/dkoshkin/kubectl-assistant/pkg/exec"
)

const (
	welcomeText = `Begin by typing what you want to accomplish in your Kubernetes cluster and then hit "Enter".
For example:
  List all control-plane Nodes
  Get Kubernetes versions for all Nodes
  Create deployment named nginx, using image nginx and ports 80 and 443
  Find all objects with label app=nginx

You will then see some text output and in most cases either a exec command or some YAML output.
If the command looks reasonable to you, type in "k" and then hit "Enter" to execute it against the cluster.

You can also type "kubectl ..." to execute a custom command.

Hit CTRL+C to exit.
`
)

type Prompter struct {
	assistantRunner assistant.Runner
	execRunner      exec.Runner

	in  io.Reader
	out io.Writer
}

func New(
	assistantRunner assistant.Runner,
	execRunner exec.Runner,
	in io.Reader,
	out io.Writer,
) Prompter {
	return Prompter{
		assistantRunner: assistantRunner,
		execRunner:      execRunner,
		in:              in,
		out:             out,
	}
}

func (p *Prompter) Loop() error {
	fmt.Fprint(p.out, welcomeText)

	ctx := context.Background()
	reader := bufio.NewReader(p.in)

	// continue until user exits
	for {
		fmt.Fprint(p.out, "\n> ")

		// ReadString will block until a newline character
		prompt, err := reader.ReadString('\n')
		if err != nil {
			return fmt.Errorf("an error occurred reading input: %w", err)
		}
		prompt = strings.TrimSpace(strings.ReplaceAll(prompt, "\n", ""))

		// convert if else to switch
		switch {
		case p.execRunner.IsKubectlCommand(prompt): // run exec commands directly if provided by user
			err = p.execRunner.Run(prompt)
			if err != nil {
				fmt.Fprintln(
					p.out,
					"An error occurred running exec command. Please try again",
					err,
				)
			}
		case prompt == "k": // ignore if a user just hits "ENTER"
			if !p.assistantRunner.CanRunKubectlCommand() {
				fmt.Fprintln(
					p.out,
					"Previous output did not contain any commands to run. Please try again",
				)
				continue
			}
			err = p.assistantRunner.RunKubectlCommand()
			if err != nil {
				fmt.Fprintln(p.out, "An error occurred running command. Please try again", err)
				continue
			}
			continue
		case prompt == "": // ignore if a user just hits "ENTER"
			continue
		default:
			// get the response from the assistant
			resp, err := p.assistantRunner.GetResponse(ctx, prompt)
			if err != nil {
				fmt.Fprintln(p.out, "An error occurred generating output. Please try again", err)
				continue
			}

			// print a separator and then the response from the assistant
			fmt.Fprintln(p.out, strings.Repeat("=", terminalWidth()))
			fmt.Fprintln(p.out, resp)
		}
	}
}

func terminalWidth() int {
	width, _, err := term.GetSize(0)
	if err != nil {
		width = 100
	}
	return width
}
